### ISSUE TITLE
[NEWS] Nowa wersja zasobów /sale/product-offers - asynchroniczne API / A new version of /sale/product-offers resources - asynchronous API

### ISSUE THREAD

**Issue by PrzemyslawLukanowski** [allegro/allegro-api#3865]
Od dziś możesz skorzystać z wersji **beta.v2** zasobów:

- [POST /sale/product-offers](https://developer.allegro.pl/documentation/#operation/createProductOffers),
- [PATCH /sale/product-offers/{offerId}](https://developer.allegro.pl/documentation/#operation/editProductOffers).

W wersji **beta.v2** wdrożyliśmy wzorzec **asynchronicznego API** ze względu na dłuższy czas wykonania niektórych operacji po stronie platformy Allegro, takich jak np. zmiany statusu publikacji oferty.

**Ważne! Wersję beta.v1 wyłączymy 12 grudnia 2020 roku.**

Aby skorzystać z wersji **beta.v2** wystarczy, że przekażesz w nagłówkach **Content-Type** oraz **Accept** wartość “**application/vnd.allegro.beta.v2+json**”.

W wersji **beta.v2** w odpowiedzi na prawidłowe żądanie na zasobie **/sale/product-offers** otrzymasz jeden z trzech statusów:

- **200 OK** - zmiany wdrożymy od razu. Występuje tylko w przypadku metody **PATCH**,
- **201 Created** - ofertę utworzymy od razu. Występuje tylko w przypadku metody **POST**,
- **202 Accepted** - ze względu na dłuższy czas wykonania operacji zadanie przeprowadzimy asynchronicznie. Występuje w przypadku metody **POST** i **PATCH**.

Wraz ze statusem **202 Accepted** zwrócimy w odpowiedzi nagłówek **Location**, w którym znajdziesz odnośnik do nowego zasobu w postaci: **/sale/product-offers/{offerId}/operations/{operationId}**. Odpytaj go metodą **GET**, aby sprawdzić status zadania. W odpowiedzi zwrócimy jeden z dwóch statusów:

- **202 Accepted** - operacja nie została jeszcze zakończona. Powtórz poprzednie żądanie.
- **303 See Othe**r - operacja została zakończona. W nagłówku **Location** przekażemy odnośnik do zasobu w postaci: /sale/product-offers/{offerId}. Skorzystaj z metody GET i przesłanego odnośnika w polu Location. W odpowiedzi zwrócimy aktualne dane oferty.

Więcej informacji o nowym sposobie działania zasobów /sale/product-offers znajdziesz w naszym poradniku:

- [publikacja oferty w asynchronicznym API](https://developer.allegro.pl/product_offers/#publikacja-oferty-w-asynchronicznym-api),
- [edycja oferty w asynchronicznym API](https://developer.allegro.pl/product_offers/#edycja-oferty-w-asynchronicznym-api).

-----

From today you can use **beta.v2** version of the following resources:

- [POST /sale/product-offers](https://developer.allegro.pl/documentation/#operation/createProductOffers),
- [PATCH /sale/product-offers/{offerId}](https://developer.allegro.pl/documentation/#operation/editProductOffers).

In the **beta.v2** version, we implemented the **asynchronous API** pattern due to the longer duration of some operations on the Allegro platform, such as e.g. changes in the publication status of the offer.

**Note! We will disable the beta.v1 version on December 12th, 2020.**

Change the data in **Content-Type** and **Accept** headers to “**application/vnd.allegro.beta.v2+json**” to start using **beta.v2** version.

You will receive one of three statuses in **beta.v2**, in response to a valid request to the sale/product-offers:

- **200 OK** - we will implement the changes immediately. Only occurs with the **PATCH** method,
- **201 Created** - we will create an offer immediately. Only occurs with the **POST** method,
- **202 Accepted** - we will perform the task asynchronously due to the longer duration of the operation. Occurs with the **POST** and **PATCH** methods.

Along with the **202 Accepted** status, we will return the **Location** header in the response, in which you will find a link to the new resource - **/sale/product-offers/{offerId}/operations/{operationId}**. Use the **GET** method on it to check the task status. In response, you will receive one of two statuses:

- **202 Accepted** - the operation has not been completed yet. Repeat the previous request.
- **303 See Other** - the operation is finished. In the **Location** header, we will return a link to the new resource - **/sale/product-offers/{offerId}**. Use the GET method on it and in the response you will receive the current offer data.

You can find more information about the new version of /sale/product-offers resources in our guide:

- [offer publication in asynchronous API](https://developer.allegro.pl/en/product_offers/#offer-publication-in-asynchronous-api),
- [editing an offer in the asynchronous API](https://developer.allegro.pl/en/product_offers/#editing-an-offer-in-the-asynchronous-api).

**Comment by Onixarts** [comment]
Rozumiem, że nie ma problemu, aby w odpowiedzi 303 po edycji oferty wywołać na niej GET /sale/offers/{offerId} zamiast proponowanego w opisie GET /sale/product-offers/{offerId}?

Chodzi o to, że potrzebujemy pełnych danych oferty, wraz z np. parametrami specyficznymi dla kategorii/produktu a product-offers tego nie zwraca.

**Comment by PrzemyslawLukanowski** [comment]
Tak, w takiej sytuacji również otrzymasz aktualne dane oferty.

**Comment by dlis2010** [comment]
Witam,

W którym momencie następuje teraz walidacja danych?
Czy przy opcji asynchronicznej i otrzymaniu 202 Accepted  może jeszcze coś pójść nie tak?

**Comment by PrzemyslawLukanowski** [comment]
@dlis2010 Nie, po 202 możesz otrzymać jedynie 303 lub ponownie 202. Informację o błędach zwrócimy już na samym początku.

**Comment by PrzemyslawLukanowski** [comment]
@dlis2010 Sprostuję - mogą się zdarzyć sytuacje, że po 202 otrzymasz 422, np. po przekroczeniu maksymalnej liczby aktywnych ofert u sprzedawcy.

**Comment by vitapak** [comment]
"202 Accepted - ze względu na dłuższy czas wykonania operacji zadanie przeprowadzimy asynchronicznie"

W jaki sposób mogę przetestować zwrotkę 202? Poproszę o przykładowy ean/kategorię, bądź sugestię, które przesłane parametry wywołają 202.

**Comment by AureliuszBrussy** [comment]
Aktualnie odpowiedź o statusie 202 wywołują requesty dotyczące zmiany statusu publikacji oferty. Wystarczy, że wystawisz nową ofertę korzystając z **POST /sale/product-offers** lub zmienisz status istniejącej oferty za pomocą **PATCH /sale/product-offers/{offerId}**. W takich przypadkach istnieje duże prawdopodobieństwo, że zwrócimy odpowiedź ze statusem 202.
